pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -251,22 +269,256 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
+	if !ok {
+		pl.logger.Error(nil, "unexpected object in isSchedulableAfterPodSchedulingChange", "obj", obj)
+		return framework.QueueAfterBackoff
+	}
+
+	// If the drivers have provided information about all
+	// unallocated claims with delayed allocation, then the next
+	// scheduling attempt is able to pick a node, so we let it run
+	// immediately if this occurred for the first time, otherwise
+	// we allow backoff.
+	pendingDelayedClaims := 0
+	if err := pl.foreachPodResourceClaim(pod, func(podResourceName string, claim *resourcev1alpha2.ResourceClaim) {
+		if claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer &&
+			claim.Status.Allocation == nil &&
+			!podSchedulingHasClaimInfo(podScheduling, podResourceName) {
+			pendingDelayedClaims++
+		}
+	}); err != nil {
+		// This is not an unexpected error: we know that
+		// foreachPodResourceClaim only returns errors for "not
+		// schedulable".
+		pl.logger.V(4).Info("pod is not schedulable, keep waiting", "pod", klog.KObj(pod), "reason", err.Error())
+		return framework.QueueSkip
+	}
+
+	if pendingDelayedClaims > 0 {
+		// We could start a pod scheduling attempt to refresh the
+		// potential nodes list.  But pod scheduling attempts are
+		// expensive and doing them too often causes the pod to enter
+		// backoff. Let's wait instead for all drivers to reply.
+		if loggerV := pl.logger.V(6); loggerV.Enabled() {
+			loggerV.Info("PodSchedulingContext with missing resource claim information, keep waiting", "pod", klog.KObj(pod), "podSchedulingDiff", cmp.Diff(oldPodScheduling, podScheduling))
+		} else {
+			pl.logger.V(5).Info("PodSchedulingContext with missing resource claim information, keep waiting", "pod", klog.KObj(pod))
+		}
+		return framework.QueueSkip
+	}
+
+	if oldPodScheduling == nil || len(oldPodScheduling.Status.ResourceClaims) < len(podScheduling.Status.ResourceClaims) {
+		// This definitely is new information for the scheduler. Try again immediately.
+		pl.logger.V(4).Info("PodSchedulingContext for pod got completed, schedule immediately", "pod", klog.KObj(pod))
+		return framework.QueueImmediately
+	}
+
+	// The other situation where the scheduler needs to do
+	// something immediately is when the selected node doesn't
+	// work: waiting in the backoff queue only helps eventually
+	// resources on the selected node become available again. It's
+	// much more likely, in particular when trying to fill up the
+	// cluster, that the choice simply didn't work out. The risk
+	// here is that in a situation where the cluster really is
+	// full, backoff won't be used because the scheduler keeps
+	// trying different nodes. This should not happen when it has
+	// full knowledge about resource availability (=
+	// PodSchedulingContext.*.UnsuitableNodes is complete) but may happen
+	// when it doesn't (= PodSchedulingContext.*.UnsuitableNodes had to be
+	// truncated).
+	if podScheduling.Spec.SelectedNode != "" {
```

Comment:
I'd like to understand this part correctly.

Let's say Node1 - 4 are in the cluster and Node2-4 doesn't have enough capacity. So, the scheduler set `podScheduling.Spec.SelectedNode=Node1`. But, later, the device driver said `claimStatus.UnsuitableNodes = Node1`. And we reach here and the Pod will get QueueImmediately.

In this case, in the next scheduling cycle, Node2-4 are still rejected by the resource fit plugin or whatever, and Node1 is rejected by the DRA plugin. Meaning, this Pod won't be schedulable until Node2-4 get free space.
But, the Pod could keep getting rescheduled by QueueImmediately.

You mentioned such risk in the above comment, right? 
How can we avoid this situation? Could you elaborate `This should not happen when it has full knowledge about resource availability`?

Reply:
> But, the Pod could keep getting rescheduled by QueueImmediately.

There will only be one scheduling attempt, the one triggered by this cluster event. After that, there won't be any further updates of the `PodSchedulingContext` unless a driver reports a change in resource availability, so there will be no further "immediate" queueing.

> You mentioned such risk in the above comment, right?

That risk is slightly different: `UnsuitableNodes` has a size limit. If a driver cannot report all unsuitable nodes, then the scheduler might alternate between different "selected nodes" that are all unsuitable.

This is an issue that we need to keep an eye on, but I think it can be ignored for now in this PR (alpha stage, not limited to scheduler events).

Reply:
> There will only be one scheduling attempt, the one triggered by this cluster event. 

Ah, true.

> UnsuitableNodes has a size limit. If a driver cannot report all unsuitable nodes, then the scheduler might alternate between different "selected nodes" that are all unsuitable.

OK, that's what you said "truncated". Understood.

Reply:
Can we have a comment to describe the risk is acceptable for now?

Reply:
Added. I also added a few more comments in other places.